File: run_LCV.m
function [ zsc_asym,gcp_est,gcp_err,rho,rho_err,likelihood,p_fullcausal1,p_fullcausal2,asym_jk1,asym_jk2,s,s_err,intercept] = run_LCV( ell,Z1,Z2,crosstrait_intercept,ldsc_intercept,weights,sig_threshold,...
    no_blocks,cross_int,n1,n2)
%RUN_LCV runs LCV on summary statistics for two traits.
%   INPUT VARIBLES: ell, Mx1 vector of LD scores; Z1, Mx1 vector of
%   estimated marginal per-normalized-genotype effects on trait 1
%   (or Z scores; invariant to scaling); Z2, Mx2 vector of effects on trait
%   2; crosstrait_intercept, 0 if cohorts are disjoint, 1 if cohorts are
%   possibly nondisjoint and necessary correction is unknown, 2 if cohorts
%   are nondisjoint with known overlap and phenotypic covariance;
%   ldsc_intercept, 0 if intercept should be fixed and 1 otherwise;
%   weights, Mx1 vector of regression weights; sig_threshold, threshold
%   above which to discard chisq statistics for the purpose of estimating
%   the LDSC intercept if they are above sig_threshold*mean_chisq;
%   no_blocks, number of jackknife blocks.
%   OPTIONAL INPUT VARIBLES:  n1,
%   1/var(Z1), only needed if ldsc_intercept=1; n2, 1/var(Z2); noisecorr,
%   covariance between sampling errors for Z1 and Z2, only needed if
%   crosstrait_intercept=2 (if zero, equivalent to setting
%   crosstrait_intercept=0).
%   OUTPUT VARIABLES: zsc_asym, Z score for partial genetic causality;
%   gcp_est, posterior mean gcp; gcp_err, posterior standard dev; rho, \
%   estimated genetic correlation; rho_err, standard error of rho estimate;
%   CI_pval, approx likelihood for gcp on [-1:.01:1]; p_fullcausal1,
%   p-value for null that gcp=-1; p_fullcausal2, p-value for null that
%   gcp=1; k41, estimate of E(alpha1^3 alpha2); k42, estimate of
%   E(alpha2^3 alpha1); s, 1x2 vector of normalizations used for Z1, Z2,
%   proportional to sqrt(h2g); s_err, standard err of s; intercept,
%   1x3 vector containing trait 1 LDSC intercept, trait 2 LDSC intercept
%   and crosstrait intercept respectively.


grid=-1:.01:1;

%% Estimate mixed 4th moments for each jackknife block
intercept_jk=zeros(no_blocks,1);asym_jk1=zeros(no_blocks,1);asym_jk2=asym_jk1;rho_jk=asym_jk1;s1jk=rho_jk;s2jk=rho_jk;
blocksize=floor(length(Z1)/no_blocks);intercept1_jk=rho_jk;intercept2_jk=rho_jk;

if ldsc_intercept
    for jk=1:no_blocks
        ind=[1:(jk-1)*blocksize, jk*blocksize+1:length(Z1)];
        [ rho_jk(jk),asym_jk1(jk),asym_jk2(jk),intercept_jk(jk),s1jk(jk),s2jk(jk),intercept1_jk(jk),intercept2_jk(jk)] = ...
            estimate_k4( ell(ind,:),Z1(ind),Z2(ind),crosstrait_intercept,ldsc_intercept,weights(ind),sig_threshold  );
        
    end
else
    for jk=1:no_blocks
        ind=[1:(jk-1)*blocksize, jk*blocksize+1:length(Z1)];
        [ rho_jk(jk),asym_jk1(jk),asym_jk2(jk),intercept_jk(jk),s1jk(jk),s2jk(jk),intercept1_jk(jk),intercept2_jk(jk)] = ...
            estimate_k4( ell(ind,:),Z1(ind),Z2(ind),crosstrait_intercept,ldsc_intercept,weights(ind),sig_threshold,n1,n2,cross_int  );
    end
end
intercept=[mean(intercept1_jk) mean(intercept2_jk) mean(intercept_jk)];

rho=mean((rho_jk));
rho_err=std((rho_jk))*sqrt(no_blocks+1);

s=[mean((s1jk)), mean((s2jk))];
s_err=[std((s1jk)), std((s2jk))]*sqrt(no_blocks+1);

asym_jk1=asym_jk1-3*rho_jk;
asym_jk2=asym_jk2-3*rho_jk;


%% Point estimate + CI
likelihood=zeros(1,length(grid));
for kk=1:length(grid) % Loop over possible gcp values
    xx=grid(kk);
    fx=abs(rho_jk).^(xx);
    numer=asym_jk2./fx-fx.*asym_jk1;
    denom=max(1./abs(rho_jk),sqrt(asym_jk2.^2./fx.^2+fx.^2.*asym_jk1.^2));
    pct_diff_jk=numer./denom;% S(xx) statistic for each jackknife block
    est_err=std(pct_diff_jk)*sqrt(no_blocks+1); % std err of S(xx)
    
    likelihood(kk)=tpdf(real(mean(pct_diff_jk)/est_err),no_blocks-2);
    if kk==1 % test for gcp=-1
        p_fullcausal1=tcdf(real(mean(pct_diff_jk)/est_err),no_blocks-2);
    elseif kk==length(grid) % test for gcp=1
        p_fullcausal2=tcdf(real(mean(pct_diff_jk)/est_err),no_blocks-2,'upper');
    elseif xx==0 % test for gcp=0
        zsc_asym=mean(pct_diff_jk)/est_err;
    end
end

gcp_est=sum(likelihood.*grid)./sum(likelihood); % posterior mean
gcp_err=sqrt(sum(likelihood.*grid.^2)./sum(likelihood)-gcp_est.^2); % posterior err

% Warnings for poorly significant or negative h2g estimates
if ~isreal(rho)
    warning('Negative heritability estimates leading to unstable results and false positives')
elseif any(s./s_err<4)
    warning('Very noisy heritability estimates potentially leading to false positives')
elseif any(s./s_err<7)
    warning('Borderline noisy heritability estimates potentially leading to false positives')
end

% Warning for non-significant rho_g
if abs(rho/rho_err)<2
    warning('No significantly nonzero genetic correlation, potentially leading to conservative p-values and difficulties in interpretation')
end

end

File: R/RunLCV.R
# RunLCV runs LCV on summary statistics for two traits.
#	Input vectors are sorted lists of LD scores and signed summary statistics. They
#		must be sorted by genomic position, as LCV uses a block-jackknife procedure
#		to compute standard errors; if consecutive SNPs are not approximately
#		contiguous, standard errors will be underestimated.
#
#   INPUT VARIBLES: 
#	ell, Mx1 vector of LD scores; 
#	z.1, Mx1 vector of estimated marginal per-normalized-genotype effects on trait 1
#   	(or Z scores; invariant to scaling);  
#	z.2, Mx2 vector of effects on trait2;  
#   no.blocks, number of jackknife blocks;
#	crosstrait.intercept, 0 if crosstrait LDSC intercept should be fixed and 1 otherwise;
#   ldsc.intercept, 0 if LDSC intercept should be fixed and 1 otherwise (1 is recommended);
#   weights, Mx1 vector of regression weights; 
#	sig.threshold, threshold above which to discard chisq statistics for the purpose of estimating
#   	the LDSC intercept; large-effect SNPs discarded above sig_threshold*mean(z.x^2);
#   n1, sample size for trait 1, only needed if ldsc.intercept=1; 
#   n2, sample size for trait 2, only needed if ldsc.intercept=1; 
#	intercept.12, covariance between sampling errors for Z1 and Z2, only needed if
#   	crosstrait_intercept=0. If the 2 GWAS are disjoint, this can be set to zero.
#
#   OUTPUT VARIABLES: 
#	lcv.output, a list with named entries:
#   "zscore", Z score for partial genetic causality. zscore>>0 implies gcp>0.
#	pval.gcpzero.2tailed, 2-tailed p-value for null that gcp=0.
#   "gcp.pm", posterior mean gcp (gcp=1: trait 1 -> trait 2; gcp=-1: trait 2-> trait 1); 
#   "gcp.pse", posterior standard error for gcp; 
#   "rho.est", estimated genetic correlation; 
#   "rho.err", standard error of rho estimate;
#   "pval.fullycausal" [2 entries], p-values for null that gcp=1 or that gcp=-1, respectively; 
#   "h2.zscore" [2 entries], z scores for trait 1 and trait 2 being heritable, respectively;
#   	we recommend reporting results for h2.zscore > 7 (a very stringent threshold).

RunLCV <- function(ell,z.1,z.2,no.blocks=100,crosstrait.intercept=1,ldsc.intercept=1,weights=1/pmax(1,ell),sig.threshold=.Machine$integer.max,n.1=1,n.2=1,intercept.12=0){
  mm=length(ell)
  if(length(z.1)!=mm||length(z.2)!=mm){
    stop('LD scores and summary statistics should have the same length')
  }
  
  source("MomentFunctions.R")
  grid<- (-100:100)/100;
  
  # Jackknife estimates of moments
  size.blocks=floor(mm/no.blocks)
  jackknife=matrix(0,no.blocks,8)
  for(jk in 1:no.blocks){
    if(jk==1)
    {ind<-(size.blocks+1):mm}
    else if(jk==no.blocks)
    {ind <- 1:(size.blocks*(jk-1))}
    else
    {ind<-c(1:((jk-1)*size.blocks), (jk*size.blocks+1):mm)}
    jackknife[jk,] <- EstimateK4(ell[ind],z.1[ind],z.2[ind],crosstrait.intercept,ldsc.intercept,weights[ind],sig.threshold,n.1,n.2,intercept.12,8)
    if(any(is.nan(jackknife))){
      stop('NaNs produced, probably due to negative heritability estimates. Check that summary statistics and LD scores are ordered correctly.')
    }
  }
  rho.est<-mean(jackknife[,1])
  rho.err=sd(jackknife[,1])*sqrt(no.blocks+1)
  flip=sign(rho.est)
  
  jackknife[,2:3]<-jackknife[,2:3]-3*jackknife[,1]
  
  # Likelihood of each gcp value
  gcp.likelihood=grid;gcp.likelihood[]=0
  for(kk in 1:length(grid)){
    xx<-grid[kk]
    fx<-abs(jackknife[,1])^(-xx)
    numer<-jackknife[,2]/fx-fx*jackknife[,3]
    denom=pmax(1/abs(jackknife[,1]),sqrt(jackknife[,2]^2/fx^2 + jackknife[,3]^2*fx^2 ))
    pct.diff<-numer/denom
    
    gcp.likelihood[kk]<-dt(mean(pct.diff)/sd(pct.diff)/sqrt(no.blocks+1),no.blocks-2)
    
    if(xx==-1){
      pval.fullycausal.2<-pt(-flip*mean(pct.diff)/sd(pct.diff)/sqrt(no.blocks+1),no.blocks-2)
    }
    if(xx==1){
      pval.fullycausal.1<-pt(flip*mean(pct.diff)/sd(pct.diff)/sqrt(no.blocks+1),no.blocks-2)
    }
    if(xx==0){
      zscore<- flip*mean(pct.diff)/sd(pct.diff)/sqrt(no.blocks+1)
    }
    
  }
  
  pval.gcpzero.2tailed=pt(-abs(zscore),no.blocks-1)*2
  
  gcp.pm<-WeightedMean(grid,gcp.likelihood)
  gcp.pse<-sqrt(WeightedMean(grid^2,gcp.likelihood)-gcp.pm^2)
  
  h2.zscore.1<-mean(jackknife[,5])/sd(jackknife[,5])/sqrt(no.blocks+1)
  h2.zscore.2<-mean(jackknife[,6])/sd(jackknife[,6])/sqrt(no.blocks+1)
  
  if(h2.zscore.1<4 || h2.zscore.2<4){
    warning('Very noisy heritability estimates potentially leading to false positives')
  }
  else{
    if(h2.zscore.1<7 || h2.zscore.2<7){
      warning('Borderline noisy heritability estimates potentially leading to false positives')
    }
  }
  if(abs(rho.est/rho.err)<2){
    warning('No significantly nonzero genetic correlation, potentially leading to conservative p-values')
  }
  
  lcv.output<-list(zscore=zscore,pval.gcpzero.2tailed=pval.gcpzero.2tailed,gcp.pm=gcp.pm,gcp.pse=gcp.pse,rho.est=rho.est,rho.err=rho.err,
                         pval.fullycausal=c(pval.fullycausal.1,pval.fullycausal.2),h2.zscore=c(h2.zscore.1,h2.zscore.2))
  
  return(lcv.output)
  
  
  
  
}